### PR TITLE
feat: Add support for OpenApi Specification as input

### PR DESCRIPTION
**Pull request by wtrocki**
## Investigation 

Add support for OpenApi Specification using APicurito models and OpenApi-To-GraphQL lib

## How this works

1. It picks open API file and transforms it to GraphQL Schema.
2. It picks up all the methods and let people to edit them 
3. It will rename OpenApi file so it will no longer be processed. (source)

# Tasks

- [x] oapi command line
- [x] Rest gateway support (requires) extra 
- [x] Directives support
- [x] Automatic filtering for custom fields
- [x] json and yaml support
- [x] Investigate having backend in REST along with GraphQL resolvers utilizing the same db/rest layer. 
- [x] Wrap integration into graphql module
- [x] Config enhancements
- [x] Docs

## Demo video: 
https://www.youtube.com/watch?v=mRZ585xjFf4&feature=youtu.be

## Verification

```
graphback init openapi
cd openapi/model
wget https://gist.github.com/wtrocki/2ece5c69bc585d36d601b0ce701103f7\
cd .. 
graphback generate
graphback db

```

## Status update

Working with openapi team to see if we can get this into our workflow and make sure that is performant. Current approach is simply generating separate graphql schema file that will also have custom resolvers. This can be later edited by users that can integrate schema. 

Second phase was to check if we can rely on the openapi for the resolvers generation.
Additionally there is a lot of challenges if we still want to process things by relying on the OpenAPi definition. I believe that we can have separate resolvers layer on our own.

https://github.com/IBM/openapi-to-graphql/issues/240

## Discussion

From my point we can integrate tooling inside our cli, but we might also simply leave this command as it is here with some additional tooling for wrapping stuff. 

Considering current limitations:
- No support for directive mapping
- Very limited schema 
- Duplication of the CRUD methods comming from our GraphQL API generator

I think it is reasonable for us to let people manually review generated graphql schema from OpenAPi Schema and no longer use OpenAPi schema as input. 

Additionally, we could have OpenApi Schema that will be used to contact backend - We can use OpenAPi to generate client that will be available in resolvers and mount it into context. 
However, developers will need to manually call the API inside the resolvers.

Generating fully functional GraphQL Api from just OpenAPI is risky business - it will be useful only for demo use cases and it can have serious impact on production use cases - having black box api/db layer is basically why so many graphql libraries fail to grow. 

### OpenAPI to GraphQL 

Very good and proven library that generates GraphQL schema and resolvers. 
Schema supports various and complex api definitions - library is very roboust in this sense.
Resolver layer is kinda complex but there is option to provide a custom one. 
Currently, we are using this library to output actual schema files that can be edited by users.
The reason for that is custom methods can overlap with CRUD ones. 

### Directives support

I have done quick spike to attach directives to the OpenAPi definition and it works like dream however it will be an custom format that people will need to know. 
If we settle with manual edits of the schema then users can get a better experience by supplying directives on generated schema rather than file

## Multiple api support

Currently command can process single api but we can move towards multiple api that is mapped to multiple graphql files. Opinions?

## New Architecture

I think we need to come up with new architecture that maps all the moving parts of the system

- Input sources (GraphQL Schema, OpenAPI, JSONSchema etc.)
- Modules (GraphQL module per each model file for separation of concerns) 
https://graphql-modules.com/
- Abstraction for data access logic. (DataServices type that can be DB(many subtypes) and REST
- Middleware type that can be REST, GraphQL (also with different subtypes)
- Tooling - Code Generation, validation, custom logic orchestration
- Transform templates to reference implementation (instead of many templates we can have single reference/opiniated server that is performant and production-ready - however community
should be able to build more
- Configuration - actual stack should be question-based (using new `graphback config` and `graphback init` command)
- People should be able to continue and tweet things using configurations

## Targets 

- Have clear responsibility for every layer of the subsystem
- Reuse a lot of proven open source technologies and standards as possible
- Decouple from template creation

## Challenges

- Source code promotion (dev/prod) on runtime
- Database migrations (depending on the db spike)

### CODE REVIEW

File: packages/graphback-cli/src/commands/openapi.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { commandRoot } from '../commandRoot'
+import { generate, transformOpenApiSpec } from '../helpers'
+
+export const command = 'openapi'
+
+export const desc = 'Generate GraphQL schema and resolvers based on Open Api spec'
```
Comment (craicoverflow) [inline packages/graphback-cli/src/commands/openapi.ts]:
```suggestion
export const desc = 'Generate GraphQL schema and resolvers based on OpenAPI spec'
```

File: packages/graphback-cli/src/helpers/transformOpenApiSpec.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { readFileSync, writeFileSync} from 'fs';
+import { GlobSync } from 'glob';
+import { printSchema } from 'graphql'
+import { safeLoad } from 'js-yaml'
+import { createGraphQlSchema } from "openapi-to-graphql"
+import { logError } from '../utils';
+
+
+export const transformOpenApiSpec = async () => {
+    const models = new GlobSync('model/*.yaml', { cwd: process.cwd() })
```
[the remark this answers is not in the document]
Comment (wtrocki) [reply packages/graphback-cli/src/helpers/transformOpenApiSpec.ts]: It will be doing that. I will push change soon.

**Comment by wtrocki** [comment]
@craicoverflow Awesome review!  Thank you

**Comment by wtrocki** [comment]
Ready for another look.

File: packages/graphback-cli/src/helpers/transformOpenApiSpec.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import { GlobSync } from 'glob';
+import { printSchema } from 'graphql';
+import { safeLoad } from 'js-yaml'
+import { createGraphQlSchema } from "openapi-to-graphql"
+import { logError, logInfo } from '../utils';
+import { removeCommentsFromSchema, removeOperationsFromSchema } from "../utils/openApiHelpers"
+/**
+ * Configuration options for OpenAPI generator
+ */
+interface OpenApiConfig {
```
[the remark this answers is not in the document]
Comment (wtrocki) [reply packages/graphback-cli/src/helpers/transformOpenApiSpec.ts]: See Config PR. Plan is to merge config PR and then connect this interface.
I was waiting for @MStokluska to merge first.

File: packages/graphback-cli/src/helpers/transformOpenApiSpec.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        await processSingleDefinition(model, path, false, openApi);
+    }
+
+    for (const model of models.found) {
+        await processSingleDefinition(model, path, true, openApi);
+    }
+
+    logInfo(`
+   Successfully generated GraphQL schema from OpenAPI definition.
+   You can review your schema in model folder and OpenAPI it for your own needs.
```
[the remark this answers is not in the document]
Comment (wtrocki) [reply packages/graphback-cli/src/helpers/transformOpenApiSpec.ts]: LOL this is funny typo

File: packages/graphback-cli/src/helpers/transformOpenApiSpec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wtrocki) [reply packages/graphback-cli/src/helpers/transformOpenApiSpec.ts]: I will rebase things from config and get that connected. 
I could not do all the job here to not obfuscate results.

**Comment by wtrocki** [comment]
We need to find a way to map ID.

**Comment by wtrocki** [comment]
Ready for review.

**Comment by wtrocki** [comment]
I have decided to add docs only. Validation phase will require a lots of refactoring so going to add it in separate PR.

**Comment by wtrocki** [comment]
@ankitjena Can I have a review on this. This one is blocking the release.

**Pull request merged**